### PR TITLE
[Merged by Bors] - Clean up Breakout logic

### PR DESCRIPTION
**Pull request by alice-i-cecile**
# Objective

1. Spawning walls in the Breakout example was hard to follow and error-prone.
2. The strategy used in `paddle_movement_system` was somewhat convoluted.
3. Correctly modifying the size of the arena was hard, due to implicit coupling between the bounds and the bounds that the paddle can move in.

## Solution

1. Refactor this to use a WallBundle struct with a builder; neatly demonstrating some essential patterns along the way.
2. Use clamp and avoid using weird &mut strategies.
3. Refactor logic to allow users to tweak the brick size, and automatically adjust the number of rows and columns to match.
4. Make the brick layout more like classic breakout!

image

### CODE REVIEW

File: examples/game/breakout.rs [hunk 1]
```diff
@@ -266,11 +276,17 @@ fn paddle_movement_system(
...
+    paddle_transform.translation.x += direction * PADDLE_SPEED * TIME_STEP;
+
+    // Bound the paddle within the walls
+    // This is maximum x value that the paddle's transform can have without leaving the play area
+    let paddle_bounds =
+        (PLAY_AREA_BOUNDS.x - WALL_THICKNESS - PADDLE_SIZE.x) / 2.0 - PADDLE_PADDING;
+    paddle_transform.translation.x = paddle_transform
+        .translation
+        .x
+        .clamp(-paddle_bounds, paddle_bounds);
```
Comment (KDecay) [inline examples/game/breakout.rs:289]: I think it might be clearer to put the new paddle position into a variable, clamp it and then assign it to `paddle_tansform.translation.x`.

So something like this:

```suggestion
    // This is the maximum x value that the paddle's transform can have without leaving the play area
    let paddle_bounds =
        (PLAY_AREA_BOUNDS.x - WALL_THICKNESS - PADDLE_SIZE.x) / 2.0 - PADDLE_PADDING;

    // Calculate the new horizontal paddle position
    let mut new_paddle_position =
        paddle_transform.translation.x + direction * PADDLE_SPEED * TIME_STEP;

    // Clamp the new horizontal paddle position to be within the paddle bounds
    new_paddle_position = new_paddle_position.clamp(-paddle_bounds, paddle_bounds);

    // Update the paddle position with the new paddle position
    paddle_transform.translation.x = new_paddle_position;
```

**Review by KDecay — CHANGES_REQUESTED** [review]

File: examples/game/breakout.rs [hunk 2]
```diff
@@ -76,24 +78,84 @@ struct Collider;
...
+        // We need to convert our Vec2 into a Vec3, by giving it a z-coordinate
+        .extend(0.0);
+
+        let scale = match orientation {
+            Left => Vec2::new(WALL_THICKNESS, PLAY_AREA_BOUNDS.y + WALL_THICKNESS),
+            Right => Vec2::new(WALL_THICKNESS, PLAY_AREA_BOUNDS.y + WALL_THICKNESS),
+            Bottom => Vec2::new(PLAY_AREA_BOUNDS.x + WALL_THICKNESS, WALL_THICKNESS),
+            Top => Vec2::new(PLAY_AREA_BOUNDS.x + WALL_THICKNESS, WALL_THICKNESS),
+        }
+        .extend(1.0);
```
Comment (rparrett) [inline examples/game/breakout.rs:121]: A comment here like the one above would be great. New users often trip over z scale in 2d.

(Or maybe this example should use `custom_size` instead of scale... hm.)

File: examples/game/breakout.rs [hunk 3]
```diff
@@ -158,79 +220,26 @@ fn setup(mut commands: Commands, asset_server: Res<AssetServer>) {
...
+    // Walls
+    commands.spawn_bundle(WallBundle::new(WallLocation::Left));
+    commands.spawn_bundle(WallBundle::new(WallLocation::Right));
+    commands.spawn_bundle(WallBundle::new(WallLocation::Bottom));
+    commands.spawn_bundle(WallBundle::new(WallLocation::Top));
 
-    // Add bricks
-    let bricks_width = BRICK_COLUMNS as f32 * (BRICK_SIZE.x + BRICK_SPACING) - BRICK_SPACING;
+    // Bricks
+    let brick_width = BRICK_COLUMNS as f32 * (BRICK_SIZE.x + BRICK_SPACING) - BRICK_SPACING;
```
Comment (rparrett) [inline examples/game/breakout.rs:230]: This rename seems slightly less clear to me.

File: examples/game/breakout.rs [hunk 4]
```diff
@@ -76,24 +78,84 @@ struct Collider;
...
+        use WallLocation::*;
+
+        let translation = match orientation {
+            Left => Vec2::new(-PLAY_AREA_BOUNDS.x / 2.0, 0.0),
+            Right => Vec2::new(PLAY_AREA_BOUNDS.x / 2.0, 0.0),
+            Bottom => Vec2::new(0.0, -PLAY_AREA_BOUNDS.y / 2.0),
+            Top => Vec2::new(0.0, PLAY_AREA_BOUNDS.y / 2.0),
+        }
+        // We need to convert our Vec2 into a Vec3, by giving it a z-coordinate
+        .extend(0.0);
```
Comment (IceSentry) [inline examples/game/breakout.rs:113]: I know it's completely valid rust, but for people new to rust it just looks like a function on the match statement. I think I would call the `.extend()` directly in the struct instantiation instead.

File: examples/game/breakout.rs [hunk 5]
```diff
@@ -158,79 +220,26 @@ fn setup(mut commands: Commands, asset_server: Res<AssetServer>) {
...
-                ..default()
-            },
-            sprite: Sprite {
-                color: WALL_COLOR,
-                ..default()
-            },
-            ..default()
-        })
-        .insert(Collider);
+    // Walls
```
Comment (IceSentry) [inline examples/game/breakout.rs:223]: This is so much nicer 👍

File: examples/game/breakout.rs (hunk 3 above)
Comment (IceSentry) [reply examples/game/breakout.rs:230]: It's the width of a single brick, not all the bricks. It makes sense to me.

File: examples/game/breakout.rs (hunk 1 above)
Comment (IceSentry) [reply examples/game/breakout.rs:289]: I would skip the reassignment of `new_paddle_position` to encourage immutability where possible and just call clamp on the last line, but otherwise I agree.

File: examples/game/breakout.rs (hunk 3 above)
Comment (IceSentry) [reply examples/game/breakout.rs:230]: It's the width of a single brick, not all the bricks. It makes sense to me.

**Comment by alice-i-cecile** [comment]
Alright, review comments addressed :) Those were super helpful.

Overall, I'm feeling way better about how this example looks compared to when I'm started: I think this is now something we can happily show off to beginners!

File: examples/game/breakout.rs [hunk 6]
```diff
@@ -76,24 +78,86 @@ struct Collider;
...
+        };
+
+        WallBundle {
+            sprite_bundle: SpriteBundle {
+                transform: Transform {
+                    // We need to convert our Vec2 into a Vec3, by giving it a z-coordinate
+                    // This is used to determine the order of our sprites
+                    translation: position.extend(0.0),
+                    // The z-scale of 2D objects must always be 1.0,
+                    // or their ordering will be affected in surpirising ways.
```
Comment (rparrett) [inline examples/game/breakout.rs:127]:
```suggestion
                    // or their ordering will be affected in surprising ways.
```

**Comment by rparrett** [comment]
Should we adjust the various padding to get the game a little closer to how it was pre-PR by default?

As the game is currently configured, it's more difficult than it used to be due to the much smaller space between the paddle and the bricks.

I'm also not sure that by default, the gaps between bricks should be larger than a ball-width.

Although the extra padding at the top is pretty classic breakout.

**Comment by alice-i-cecile** [comment]
> Should we adjust the various padding to get the game a little closer to how it was pre-PR by default?

Yep, totally open to this. I don't think there was a *strong* reason to think it was optimal before, but I'm happy to fiddle with the numbers.

Definitely agree on the gap size for sure.

**Comment by cart** [comment]
>  I don't think there was a strong reason to think it was optimal before, but I'm happy to fiddle with the numbers.

I think having consistent vertical and horizontal padding definitely looks "better" to the average eye. The old settings had consistent padding on the top, left, and right, which makes it preferable imo. We should also consider making the bottom part consistent.

**Review by KDecay — APPROVED** [review]
